[JSC] Wasm: OMG B3 IR should not have a B3::Variable for every wasm expression

dhecht opened the pull request:
#### 4ad30ea67c8d320da890d364bd50cbc88a0be87e
<pre>
[JSC] Wasm: OMG B3 IR should not have a B3::Variable for every wasm expression
<a href="https://bugs.webkit.org/show_bug.cgi?id=304352">https://bugs.webkit.org/show_bug.cgi?id=304352</a>
<a href="https://rdar.apple.com/155818669">rdar://155818669</a>

Reviewed by Keith Miller and Yusuke Suzuki.

Currently, OMGIRGenerator creates a B3::Variable and corresponding
Set and Get B3::Values for every wasm stack expression. It does this
so that if a catch handler entrypoint or loop OSR entrypoint is
needed, these alternate entry points have a way to modify the wasm
state (to unwind the stack or OSR enter). Inserting Phis for this
is difficult in a single pass, and this way, we leverage the existing
fixSSA pass to convert the Get/Set/Variables into Phis.

However, this pessimizes the usual case and is wasteful since most
wasm expressions that occur in a function are not going to be on the
wasm stack when the parser encounters Try/TryTable/Loop. So doing this
creates a lot of extraneous work for fixSSA as well as a bunch of
unnecessary allocations of these Variables and Values, only to throw
them away shortly. Also, it makes the generated IR convoluted and
more difficult to read.

Instead, let&apos;s optimistically generate B3::Value for wasm expressions,
and &quot;materialize&quot; Values into Variables only when necessary
for Values that are on the wasm stack when the parser encounters
Try/TryTable/Loop. To do this, we go back and rewrite the code a bit to
insert Sets at the points where these Values were generated. That way,
we still leverage fixSSA for Phi placement for these alternate
entrypoints, yet the vast majority of wasm stack expressions can be
SSA B3::Values from the start.

Testing: covered by existing wasm tests
Canonical link: <a href="https://commits.webkit.org/304637@main">https://commits.webkit.org/304637@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/03a5325df0394bf38dad92a490220023b1e5788d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136195 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8551 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/138/builds/47475 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/143904 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/89163 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/155/builds/9226 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/156/builds/8396 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/143904 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/89163 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139141 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/155/builds/9226 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/138/builds/47475 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/143904 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/155/builds/9226 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/138/builds/47475 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo-libwebrtc~~](https://ews-build.webkit.org/#/builders/166/builds/4499 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/128153 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/155/builds/9226 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/138/builds/47475 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/146650 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/134680 "Built successfully and passed tests") | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/153/builds/8235 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/138/builds/47475 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/146650 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/160/builds/8251 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/156/builds/8396 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/146650 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-26-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/138/builds/47475 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/62222 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20979 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/158/builds/8283 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/138/builds/47475 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/167459 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/8001 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/71842 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/167459 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/8222 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/8075 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
```diff
@@ -5032,6 +5074,76 @@ RefPtr<PatchpointExceptionHandle> OMGIRGenerator::preparePatchpointForExceptions
+        void convertToVariable(OMGExpression& expr, Variable* variable)
+        {
+            if (expr.isMaterialized() || expr.b3Value()->isConstant())
+                return; // XXX should we track for each control block?
```
Is this meant to be a FIXME?

dhecht replied:
Oops, no it was an early note to self but I don't think it turns out to be an improvement so I'll just delete.

Constellation commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
```diff
@@ -118,10 +118,76 @@ static constexpr bool traceExecutionIncludesConstructionSite = false;
+// so that the catch or loop OSR entrypoints have a location to restore into without needing
+// to determine Phi placement during this single pass parsing. Phi placement for these variables
+// will be handled by fixSSA.
+class OMGExpression {
+    static constexpr uintptr_t isMaterializedMask = 0x1;
+public:
```
This `OMGExpression` is copyable value. Is it possible that some value is marked as materialized, but other is not marked yet?

``
OMGExpression a = ...;

// Due to some reason, we copy it.
auto b = a;

// And `b` gets materialized.

...

// But use `a`.
```

dhecht replied:
Agree it should be non-copyable, but will do a follow up patch for this.

kmiller68 commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
```diff
@@ -118,10 +118,76 @@ static constexpr bool traceExecutionIncludesConstructionSite = false;
+class OMGExpression {
+    static constexpr uintptr_t isMaterializedMask = 0x1;
+public:
+    OMGExpression() = default;
```
I think this should be marked `WTF_MAKE_NONCOPYABLE` since it seems like it would be bad to duplicate this.

dhecht replied:
Agree, but will do a follow up patch for this.

kmiller68 commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
```diff
@@ -5032,6 +5074,76 @@ RefPtr<PatchpointExceptionHandle> OMGIRGenerator::preparePatchpointForExceptions
+            endBlock(nullptr);
+        }
+
+        void convertToVariable(OMGExpression& expr, Variable* variable)
```
I was wondering how this worked for loops but I realized it "just does" because you can't TryTable and Try-Catch can't catch backwards to the loop header. i.e. the label you catch at is always after the throw point in the wasm bytecode and structured IR. Might be worth a comment here explaining that.

dhecht replied:
Sure, will add a comment.

kmiller68 reviewed (approved):
Nice! r=me

dhecht committed:
[JSC] Wasm: OMG B3 IR should not have a B3::Variable for every wasm expression
https://bugs.webkit.org/show_bug.cgi?id=304352
rdar://155818669

Reviewed by Keith Miller and Yusuke Suzuki.

Currently, OMGIRGenerator creates a B3::Variable and corresponding
Set and Get B3::Values for every wasm stack expression. It does this
so that if a catch handler entrypoint or loop OSR entrypoint is
needed, these alternate entry points have a way to modify the wasm
state (to unwind the stack or OSR enter). Inserting Phis for this
is difficult in a single pass, and this way, we leverage the existing
fixSSA pass to convert the Get/Set/Variables into Phis.

However, this pessimizes the usual case and is wasteful since most
wasm expressions that occur in a function are not going to be on the
wasm stack when the parser encounters Try/TryTable/Loop. So doing this
creates a lot of extraneous work for fixSSA as well as a bunch of
unnecessary allocations of these Variables and Values, only to throw
them away shortly. Also, it makes the generated IR convoluted and
more difficult to read.

Instead, let's optimistically generate B3::Value for wasm expressions,
and "materialize" Values into Variables only when necessary
for Values that are on the wasm stack when the parser encounters
Try/TryTable/Loop. To do this, we go back and rewrite the code a bit to
insert Sets at the points where these Values were generated. That way,
we still leverage fixSSA for Phi placement for these alternate
entrypoints, yet the vast majority of wasm stack expressions can be
SSA B3::Values from the start.

Testing: covered by existing wasm tests
Canonical link: https://commits.webkit.org/304637@main
